pyscf/mcscf/ucasci.py

```diff
@@ -296,13 +296,13 @@ def _finalize(self):
             ss_core = self._scf.spin_square(mocore, ovlp_ao)
             if isinstance(self.e_cas, (float, numpy.number)):
                 ss = fci.spin_op.spin_square(self.ci, self.ncas, self.nelecas,
-                                             mocas, ovlp_ao)
+                                             mocas, ovlp_ao, self.fcisolver)
                 log.note('UCASCI E = %.15g  E(CI) = %.15g  S^2 = %.7f',
                          self.e_tot, self.e_cas, ss[0]+ss_core[0])
             else:
                 for i, e in enumerate(self.e_cas):
                     ss = fci.spin_op.spin_square(self.ci[i], self.ncas, self.nelecas,
-                                                 mocas, ovlp_ao)
+                                                 mocas, ovlp_ao, self.fcisolver)
```

Comment:
`self.fcisolver.states_spin_square` ?

Reply:
@sunqm As explained in the May 16 reply, calling ``self.fcisolver.spin_square()`` or ``self.fcisolver. states_spin_square()`` will break the existing UCASCI code, this is why in my pull request I choose to keep ``fci.spin_op.spin_square`` but this method needs to get the 1/2rdms from the correct fcisolver. In addition, the computation of $\langle S^2 \rangle$ from 1/2rdms is general and independent from ``fcisolver`` implementation, which is what currently implemented in ``fci.spin_op.spin_square`` (See https://github.com/pyscf/pyscf/blob/master/pyscf/fci/spin_op.py#L112-L120). For DMRG fcisolver we can indeed specialize the computation of $\langle S^2 \rangle$ without computing and storing 1 and 2 rdms, which may be slightly faster, but the default FCI solver in pyscf does not have this specialization.

The error in the github action checks (below) is not relevant to this pull request, which should have been fixed at some time after I created this pull request. Someone should trigger the rerun of these checks.